JSTOR-Elasticsearch-with-embeddings.py

```diff
@@ -0,0 +1,540 @@
+
+            try:
+                if isinstance(contrib_group['aff'], str):
+                    contrib_group['aff'] = [contrib_group['aff']]
+            except KeyError:
+                pass
+
+            for i, contrib in enumerate(contrib_group['contrib']):
+                try:
+                    name = contrib['string-name']
+                except KeyError:
+                    pass
+                else:
+                    if isinstance(name, str):
+                        name = [name]
+                    contrib_group['contrib'][i]['string-name'] = name
+                try:
+                    if isinstance(contrib['aff'], str):
+                        contrib['aff'] = [contrib['aff']]
+                except KeyError:
+                    pass
+
+        return contrib_groups
+
+    def _process__article_meta__pub_date(self, pub_dates):
+        years = None
+        if isinstance(pub_dates, dict):
+            if isinstance(pub_dates['year'], list):
+                return pub_dates['year']
+            else:
+                return [pub_dates['year']]
+
+        return [pub_date['year'] for pub_date in pub_dates]
+
+    def _process__article_meta(self, article_meta):
+        _make_str = lambda d: d.get('#text', None) if isinstance(d, dict) else d
+
+        # Remove duplicates
+        for key in ('issue-id', 'issue', 'volume', 'pub-date'):
+            if isinstance(article_meta.get(key, ''), list):
+                if article_meta[key][0] == article_meta[key][1]:
+                    article_meta[key] = article_meta[key][0]
+
+        self._process_if_avail(article_meta, 'pub-date', self._process__article_meta__pub_date, 'year')
+        # self._process_if_avail(article_meta, 'contrib-group', self._process__article_meta__contrib_group)
+        # self._process_if_avail(article_meta, 'issue-id', _make_str)
+        # self._process_if_avail(article_meta, 'issue', _make_str)
+        # self._process_if_avail(article_meta, 'volume', _make_str)
+
+        try:
+            self._process_if_avail(article_meta['title-group'], 'article-title', _make_str)
+        except KeyError:
+            pass
+        return article_meta
+
+    def _process__journal_meta(self, journal_meta):
+        journal_title_str = lambda journal_title: journal_title['#text'] if isinstance(journal_title,
+                                                                                       dict) else journal_title
+        self._process_if_avail(journal_meta['journal-title-group'], 'journal-title', journal_title_str)
+        try:
+            journal_meta['journal-title'] = journal_meta.pop('journal-title-group')['journal-title']
+        except KeyError:
+            pass
+        return journal_meta
+
+    def _get_parse_postprocessor(self, article_et):
+        etree_str = lambda e: etree.tostring(e, encoding='utf-8', method='text').strip()
+        isenglish = lambda x: re.match(x, 'eng?', re.I)
+
+        def postprocessor(path, key, value):
```

Comment:
This is a giant method. I would try to refactor parts into smaller methods if possible. It will make your code easier to read and more reusable if the need should arise.